x/paychan/handler.go

```diff
@@ -0,0 +1,296 @@
+			Whole:      0,
+			Fractional: 0,
+			Ticker:     msg.Total.Ticker,
+		},
+	})
+	if err != nil {
+		return res, err
+	}
+
+	// Move coins from sender account and deposit total amount available on
+	// that channels account.
+	src := msg.SenderPublicKey.Address()
+	dst := paymentChannelAccount(obj.Key())
+	if err := h.cash.MoveCoins(db, src, dst, *msg.Total); err != nil {
+		return res, err
+	}
+
+	res.Data = obj.Key()
+	return res, err
+}
+
+type transferPaymentChannelHandler struct {
+	auth   x.Authenticator
+	bucket PaymentChannelBucket
+	cash   cash.Controller
+}
+
+var _ weave.Handler = (*transferPaymentChannelHandler)(nil)
+
+func (h *transferPaymentChannelHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.CheckResult, error) {
+	var res weave.CheckResult
+	if _, err := h.validate(ctx, db, tx); err != nil {
+		return res, err
+	}
+	res.GasAllocated += transferPaymentChannelCost
+	return res, nil
+}
+
+func (h *transferPaymentChannelHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*TransferPaymentChannelMsg, error) {
+	rmsg, err := tx.GetMsg()
+	if err != nil {
+		return nil, err
+	}
+	msg, ok := rmsg.(*TransferPaymentChannelMsg)
+	if !ok {
+		return nil, errors.ErrUnknownTxType(rmsg)
+	}
+
+	if err := msg.Validate(); err != nil {
+		return msg, err
+	}
+
+	// TODO(husio) validate that the payment is done for the right chain
+
+	pc, err := h.bucket.GetPaymentChannel(db, msg.Payment.ChannelId)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check signature to ensure the message was not altered.
+	raw, err := msg.Payment.Marshal()
+	if err != nil {
+		return nil, fmt.Errorf("serialize payment: %s", err)
+	}
+	if !pc.SenderPublicKey.Verify(raw, msg.Signature) {
+		return msg, ErrInvalidSignature()
+	}
+
+	if !msg.Payment.Amount.SameType(*pc.Total) {
+		return msg, ErrInvalidAmount(msg.Payment.Amount)
+	}
+
+	if msg.Payment.Amount.IsGT(*pc.Total) {
+		return msg, ErrInvalidAmount(msg.Payment.Amount)
+	}
+	// Payment is representing a cumulative amount that is to be
+	// transferred to recipients account. Because it is cumulative, every
+	// transfer request must be greater than the previous one.
+	if !msg.Payment.Amount.IsGT(*pc.Transferred) {
+		return msg, ErrInvalidAmount(msg.Payment.Amount)
+	}
+
+	return msg, nil
+}
+
+func (h *transferPaymentChannelHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.DeliverResult, error) {
+	var res weave.DeliverResult
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return res, err
+	}
+
+	pc, err := h.bucket.GetPaymentChannel(db, msg.Payment.ChannelId)
+	if err != nil {
+		return res, err
+	}
+
+	// Payment amount is total amount that should be transferred from
+	// payment channel to recipient. Deduct already transferred funds and
+	// move only the difference.
+	diff, err := msg.Payment.Amount.Add(pc.Transferred.Negative())
+	if err != nil || diff.IsZero() {
+		return res, ErrInvalidAmount(msg.Payment.Amount)
```

Comment:
👍 A  zero amount can only happen on resubmission or the original sender not creating proper increasing amounts.